### PR TITLE
Create subnet by project

### PR DESCRIPTION
**Pull request by rcarrillocruz**
##### Issue Type:
- Feature Pull Request
##### Plugin Name:

os_subnet
##### Summary:

A cloud/domain admin should be able to create a subnet on any
project it is granted on.
This change adds the 'project' parameter that accepts either
a name (admin-only) or id.
##### Example:

```
(venv) ubuntu@devstack:~$ cat test_os_subnet.yml

---
- hosts: localhost
  connection: local
  gather_facts: false
  tasks:
  - os_subnet:
      cloud: devstack-cloud_admin
      state: present
      name: openstackci-subnet
      network_name: openstackci-net
      cidr: 192.168.3.0/24
      dns_nameservers:
        - 8.8.8.8
      project: 7d517c389898420d8113d557357914eb
(venv) ubuntu@devstack:~$ ansible-playbook -i inventory test_os_subnet.yml -e "@infra_config.yml" -vvv
No config file found; using defaults

PLAYBOOK: test_os_subnet.yml ***************************************************
1 plays in test_os_subnet.yml

PLAY [localhost] ***************************************************************

TASK [os_subnet] ***************************************************************
task path: /home/ubuntu/test_os_subnet.yml:6
ESTABLISH LOCAL CONNECTION FOR USER: ubuntu
localhost EXEC /bin/sh -c '( umask 22 && mkdir -p "` echo $HOME/.ansible/tmp/ansible-tmp-1457958779.2-28797766643306 `" && echo "` echo $HOME/.ansible/tmp/ansible-tmp-1457958779.2-28797766643306 `" )'
localhost PUT /tmp/tmpsu8RfO TO /home/ubuntu/.ansible/tmp/ansible-tmp-1457958779.2-28797766643306/os_subnet
localhost EXEC /bin/sh -c 'LANG=en_US.UTF-8 LC_ALL=en_US.UTF-8 LC_MESSAGES=en_US.UTF-8 /usr/bin/env python /home/ubuntu/.ansible/tmp/ansible-tmp-1457958779.2-28797766643306/os_subnet'
changed: [localhost] => {"changed": true, "id": "d1ba731f-1332-4389-946e-e96bb1bb6d38", "invocation": {"module_args": {"allocation_pool_end": null, "allocation_pool_start": null, "api_timeout": null, "auth": null, "auth_type": null, "availability_zone": null, "cacert": null, "cert": null, "cidr": "192.168.3.0/24", "cloud": "devstack-cloud_admin", "dns_nameservers": ["8.8.8.8"], "enable_dhcp": true, "endpoint_type": "public", "gateway_ip": null, "host_routes": null, "ip_version": "4", "ipv6_address_mode": null, "ipv6_ra_mode": null, "key": null, "name": "openstackci-subnet", "network_name": "openstackci-net", "region_name": null, "state": "present", "timeout": 180, "verify": true, "wait": true}, "module_name": "os_subnet"}, "subnet": {"allocation_pools": [{"end": "192.168.3.254", "start": "192.168.3.2"}], "cidr": "192.168.3.0/24", "dns_nameservers": ["8.8.8.8"], "enable_dhcp": true, "gateway_ip": "192.168.3.1", "host_routes": [], "id": "d1ba731f-1332-4389-946e-e96bb1bb6d38", "ip_version": 4, "ipv6_address_mode": null, "ipv6_ra_mode": null, "name": "openstackci-subnet", "network_id": "82e3abb6-05e3-4a70-8de7-ae0a13bdc6d6", "subnetpool_id": null, "tenant_id": "7d517c389898420d8113d557357914eb"}}

PLAY RECAP *********************************************************************
localhost                  : ok=1    changed=1    unreachable=0    failed=0

(venv) ubuntu@devstack:~$ openstack subnet show openstackci-subnet
+-------------------+--------------------------------------+
| Field             | Value                                |
+-------------------+--------------------------------------+
| allocation_pools  | 192.168.3.2-192.168.3.254            |
| cidr              | 192.168.3.0/24                       |
| dns_nameservers   | 8.8.8.8                              |
| enable_dhcp       | True                                 |
| gateway_ip        | 192.168.3.1                          |
| host_routes       |                                      |
| id                | d1ba731f-1332-4389-946e-e96bb1bb6d38 |
| ip_version        | 4                                    |
| ipv6_address_mode | None                                 |
| ipv6_ra_mode      | None                                 |
| name              | openstackci-subnet                   |
| network_id        | 82e3abb6-05e3-4a70-8de7-ae0a13bdc6d6 |
| project_id        | 7d517c389898420d8113d557357914eb     |
| subnetpool_id     | None                                 |
+-------------------+--------------------------------------+

```

### CODE REVIEW

**Comment by Shrews** [comment]
Similar to my comments in #3236, we should do some shade version checking here.

File: cloud/openstack/os_subnet.py [hunk 1]
```diff
@@ -298,7 +313,8 @@ def main():
                                         gateway_ip=gateway_ip,
                                         dns_nameservers=dns,
                                         allocation_pools=pool,
-                                        host_routes=host_routes)
+                                        host_routes=host_routes,
+                                        tenant_id=project_id)
```
Comment (Shrews) [inline cloud/openstack/os_subnet.py]: update_subnet() does not support tenant_id.

**Comment by Shrews** [comment]
Ok, further investigation says the create_subnet() in shade already supports a project ID, so no need for the version checking. However, update_subnet() does not support this parameter, so the need_revision still stands.

**Comment by rcarrillocruz** [comment]
@Shrews could you have a look now please?

Thanks

**Comment by rcarrillocruz** [comment]
ready_for_review

**Comment by gregdek** [comment]
Thanks @rcarrillocruz. To the current maintainers, @emonty, @shrews, @juliakreger, @j2sol, @rcarrillocruz please review according to guidelines (http://docs.ansible.com/ansible/developing_modules.html#module-checklist) and comment with text 'shipit' or 'needs_revision' as appropriate.

[This message brought to you by your friendly Ansibull-bot.]

File: cloud/openstack/os_subnet.py [hunk 2]
```diff
@@ -271,7 +278,14 @@ def main():
 
     try:
         cloud = shade.openstack_cloud(**module.params)
-        subnet = cloud.get_subnet(subnet_name)
+        if project is not None:
+            project_id = cloud.get_project(project)['id']
```
Comment (Shrews) [inline cloud/openstack/os_subnet.py]: This throws an exception if the project is invalid.

`TypeError: 'NoneType' object has no attribute '__getitem__'`

Comment (rcarrillocruz) [inline cloud/openstack/os_subnet.py]: good catch

**Comment by rcarrillocruz** [comment]
ready_for_review

**Comment by gregdek** [comment]
Thanks @rcarrillocruz. To the current maintainers, @emonty, @shrews, @juliakreger, @j2sol, @rcarrillocruz please review according to guidelines (http://docs.ansible.com/ansible/developing_modules.html#module-checklist) and comment with text 'shipit' or 'needs_revision' as appropriate.

[This message brought to you by your friendly Ansibull-bot.]

**Pull request merged**